File: cFiles/cs137/notes/week11/quickSort.c
#include <stdio.h>
//pick a pivot, divide the array into parts that are less than the pivot, and greater than the pivot
//two cases: pivot is element 0
		//if x >= p, just add it to end?
		//if x < p, we have to move it to left of pivot

/*
Run Time Complexity of QuickSort:
Best Case: each partitioning yields two equal-sized partitions . best case is same as margesort log2 of n levels
Average Case is O(nlogn). p falls between 25th and 75th percentile of highest, you get 3:1 partition sizes at worst, and 1:1 partition size at best
results in log (4/3) n --> 0 nlogn for QuickSort
worst case: p is always the smallest or largest element in the array. worst case is O(n ^ 2)
4 1 2 3...pivot is at 4.
time is N + n - 1 + n -2 + n - 3.... O N^2
Back to the average case, what we wanted was to select a pivot such that at worst.. what we have been using as our pivot is a[0].
what we suggest is take 3 values from the array..first, last, middle, take the median of those 3.. this increase your chances of the middle pivot being in the
25th to 75th percitle
this is to get O n log n performance
pick the mediuan of the 3 values instead of a[0] for the pivot
p1, p2, p3
swap a[0] with the median of p1, p2, p3...
another technique is to do the median of 3 medians

space complexity of merge sort
is order N --> additional storage. this is slightly fudging the truth


*/

void swap (int a[], int i, int j){
	int temp = a[i];
	a[i] = a[j];
	a[j] = temp;
}

int partition (int a[], int n){//pivot is a[0]
	int m = 0;
	for (int i = 1; i < n; i++){
		if (a[i] < a[0]){
			m++;
			swap(a, m, i);
		}
	}
	swap(a, 0, m);
	return m;
}
void quicksort(int a[], int n){
	if (n <= 1) return;
	int m = partition(a, n);
	quicksort(a, m);
	quicksort(a + m + 1, n - m - 1);
}

int main (void){
	int a[] = {2, -10, 14, 42, 11, -7, 0, 38};
	printf("Before:\n");
	for (int i = 0; i < sizeof(a)/sizeof(a[0]); i++){
		printf("%d ", a[i]);
	}
	printf("\nAfter:\n");
	quicksort(a, sizeof(a)/sizeof(a[0]));
	for (int i = 0; i < sizeof(a)/sizeof(a[0]); i++){
		printf("%d ", a[i]);
	}
	printf("\n");
}

File: interviewPrep/InterviewPrepWinter2018/src/Sorting/QuickSort.java

package Sorting;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class QuickSort {

    // Select a pivot. Make all elements to left of pivot < pivot, all to the right > pivot
    
    public static int partition(int data[], int left, int right){
        int i = left;
        int j = right;
        int pivotVal = data[(left + right) / 2]; // get pivot in middle of list
        System.out.println("\npivotVal: " + pivotVal + ", left is: " + left + ", right is: " + right);
        
        while (i <= j){ // divide into 2 lists
            // If value in left list is smaller than the pivotVal, then
            // get the next element from the left list (keep going)
            while (data[i] < pivotVal){
                i++;
            }
            
            // If value in right list is larger than pivotVal, then
            // get the next element from the right list (keep going)
            while (data[j] > pivotVal){
                j--;
            }
            
            // if we have found a value in left list that is larger than pivotVal
            // and found a value in right list that is smaller than pivotVAl,
            // then swap them. and increase i and j.
            if (i <= j){
                System.out.print("Before swapping: ");
                print(data);
                swap(data, i , j);
                System.out.print("After swapping: ");
                print(data);
                i++;
                j--;
            }
        }
        return i; 
    }
    
    public static void swap(int data[], int a, int b){
        int temp = data[a];
        data[a] = data[b];
        data[b] = temp;
    }
    
    public static void quicksort(int data[], int left, int right){
        int index = partition(data, left, right);
        System.out.println("index is: " + index);
        if (left < index - 1){
            System.out.println("sorting left");
            quicksort(data, left, index - 1);
        } 
        
        if (index < right){
            System.out.println("sorting right");
            quicksort(data, index, right);
        }
    }
    
    public static void print(int data[]){
        for (int x : data){
            System.out.print(x + ", ");
        }
        System.out.print("\n");
    }
    
    public static void main(String[] args) {
        int data[] = {7,1,5,9,6,10,2};
        
        quicksort(data, 0 , data.length - 1);
        
        System.out.println("Final: ");
        print(data);
    }
    
}
